---
title: "Adaption of workflow to new dataset"

# output option 1: markdown for github

output: github_document

# output option 2: html

# output:
#   html_document:
#     theme: united
#     df_print: kable

date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
author:
    'Guido Schl√∂gel, Sonja Tockner'
---

```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)
```

In this document only the differences to the original workflow are described. A more detailed description of the workflow is found in the original tutorial description. The code can be found [here](https://github.com/gschloegel/Course_Fallstudie_scRNA-Seq-with-Seurat/blob/main/scripts/new_dataset_neurons_900.rmd)

## Preperation

The workflow of the Seurat tutorial is adapted to the new dataset:

[1k Brain Cells from an E18 Mouse](https://www.10xgenomics.com/resources/datasets/1-k-brain-cells-from-an-e-18-mouse-2-standard-2-1-0)

* Cells from a combined cortex, hippocampus and sub ventricular zone of an E18 mouse
*	Sequenced on Illumina HiSeq2500 with approximately 56,000 reads per cell
*	26bp read1 (16bp Chromium barcode and 10bp UMI), 98bp read2 (transcript), and 8bp I7 sample barcode

The filtered data set is unpacked into the data folder of the project.

## Loading libraries
```{r}
library(dplyr)
library(Seurat)
library(patchwork)
```

## Setup the seurat Object
The path has to be changed to the folder with the new data.

```{r}
# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "../data/neurons_900_filtered_gene_bc_matrices/filtered_gene_bc_matrices/mm10/")
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "neurons_900", min.cells = 3, min.features = 200)
pbmc
```

## Preprocessing and Quality control
In the new dataset the mitochondrial genes start with "mt-". The lower case is not recognized by the original pattern. The code was adapted to find the desired genes in the new dataset.
```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-|^mt-") # changing pattern

```

## Visualization of QC Metrices
```{r, fig.height=5, fig.width=11}
# Visualize QC metrics as a violin plot
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

```

## Visualization of feature-feature relationships by FeatureScatter
```{r, fig.height=5, fig.width=11}
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```

## Subsets with specific filter criteria can also be defined
The new dataset has a higher number of features. Therefore the filter was redefined according to the plots above.
```{r}
pbmc <- subset(pbmc, subset = nFeature_RNA > 1000 & nFeature_RNA < 7000 & percent.mt < 10) # adapted values according to the plots

```

## Normalization of the data

The normalization is performed using the standard parameters.

```{r}
pbmc <- NormalizeData(pbmc)
```

## Feature selection for the identification of highly variable features
The number of selected features remains unchanged for the new data
```{r, fig.height=5, fig.width=11}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2


```

## Scaling of the data
We use the standard options here as we just need the selected features afterwards. There is no need to scale the not-selected features.
```{r}
pbmc <- ScaleData(pbmc)
```
## Linear dimension reduction
```{r}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))

```
## Visualizing cells and features that define PCA
```{r}
# Examine and visualize PCA results in a few different ways
print(pbmc[["pca"]], dims = 1:5, nfeatures = 5)

```

```{r, fig.height=5, fig.width=11}
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")

```

```{r, fig.height=5, fig.width=11}
DimPlot(pbmc, reduction = "pca")

```

We observe that PC1 is evenly distributed while PC2 is influenced by a low number of values.
## Visualization with DimHeatmap()
As in the original Tutorial we use the heat maps to select the PCs containing the most relevant information for our dataset. We observe that the separation is not optimal and that PC1 has the biggest impact.
```{r, fig.height=5, fig.width=11}
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)
```

```{r, fig.height=10, fig.width=11}
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)

```
## Determine the dimensionality of a dataset
We use the JackStraw plot as well as the EllbowPlot the determine the number of PCs used for further analysis.
### Jack Straw Plot
```{r}
# NOTE: This process can take a long time for big datasets, comment out for expediency. More
# approximate techniques such as those implemented in ElbowPlot() can be used to reduce
# computation time
pbmc <- JackStraw(pbmc, num.replicate = 100)
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)

```
Comparison of the distribution of p-values for each PC with a unit distribution. Significant PCs show a strong enrichment of features with low p-values (solid curve above the dashed line).
```{r, fig.height=5, fig.width=11}
JackStrawPlot(pbmc, dims = 1:20)

```
### Elbow plot

```{r, fig.height=5, fig.width=11}
ElbowPlot(pbmc, ndims = 30)

```
### Select PCs
The Jack Straw Plot is not conclusive. The ellbow plot suggest that 15 PCs could be a good choice. We will later try different amount of PCs to check if the results are robust to these changes.
```{r}
used_dims = 15
```



## Cluster the cells
We use the standard graph based clustering approach as we do not know how many cluster we can expect.

```{r}
pbmc <- FindNeighbors(pbmc, dims = 1:used_dims)
pbmc <- FindClusters(pbmc, resolution = 0.5)
```

```{r}
# Look at cluster IDs of the first 5 cells
head(Idents(pbmc), 5)

```
## Non-linear dimensional reduction(UMAP)
We use UMAP to visualize the clusters. 

```{r}
pbmc <- RunUMAP(pbmc, dims = 1:used_dims)

```

```{r, fig.height=5, fig.width=11}
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(pbmc, reduction = "umap")

```
The name of the saved Seurat object is changed to seperate the different datasets.
```{r}
### output ordner angelegt
saveRDS(pbmc, file = "../output/pbmc_neurons_900.rds")

```

## Search for differentially expressed traits (cluster biomarkers)

We search for the 2 genes with the highest fold change for each cluster. Later we use them to mark the clusters.

```{r}
# find markers for every cluster compared to all remaining cells, report only the positive
# ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
biomarkers <- pbmc.markers %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)
print(biomarkers)

```

## Visualization of the marker expression
We use the VinPlot to show the distribution of the genes with the highest positive and negative loadings of PC1. We see the different expressions in the different clusters and can, for example, see that Phgdh is almost exlusivly expressted in the clusters 3 and 4.
```{r, fig.height=5, fig.width=11}
VlnPlot(pbmc, features = c("Dbi", "Phgdh"))
```

```{r, fig.height=5, fig.width=11}
# you can plot raw counts as well
VlnPlot(pbmc, features = c("Tubb3", "Tmsb10"), slot = "counts", log = TRUE)
```
Nest we check our marker genes for the different clusters. We slice the biomarkers data frame to just view the genes with the highest fold change.
```{r, fig.height=5, fig.width=11}
FeaturePlot(pbmc, features = biomarkers$gene[c(TRUE, FALSE)])
```

```{r, include=FALSE}
library(ggplot2)
plot <- FeaturePlot(pbmc, features = biomarkers$gene[c(TRUE, FALSE)]) + 
  theme(axis.title = element_text(size = 18), legend.text = element_text(size = 18)) + 
  guides(colour = guide_legend(override.aes = list(size = 10)))
ggsave(filename = "../output/neurons_900_FeaturePlot.jpg", height = 7, width = 12, plot = plot, quality = 50)
```

DoHeatmap can be used to generate an expression heatmap for cells and features. The top 20 markers for each cluster are plotted here.
```{r, fig.height=5, fig.width=11}
pbmc.markers %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC) -> top10
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```

## Assigning marker genes to the clusters
In this step we do not have enough background information to link our marker genes with cell types. Therefore we mark the clusters with the best marker gene.

    
```{r, fig.height=5, fig.width=11}
new.cluster.ids <- biomarkers$gene[c(TRUE, FALSE)]
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

```{r save.img, include=FALSE}
library(ggplot2)
plot <- DimPlot(pbmc, reduction = "umap", label = TRUE, label.size = 4.5) + xlab("UMAP 1") + ylab("UMAP 2") + 
  theme(axis.title = element_text(size = 18), legend.text = element_text(size = 18)) + 
  guides(colour = guide_legend(override.aes = list(size = 10)))
ggsave(filename = "../output/neurons_900_umap.jpg", height = 7, width = 12, plot = plot, quality = 50)
```

```{r}
saveRDS(pbmc, file = "../output/neurons_900_final.rds")
```

<details>
  <summary>**Session Info**</summary>
```{r}
sessionInfo()
```
</details>
