---
title: "Replication and Description of Seurat Tutorial"

# output option 1: markdown for github

output: github_document

# output option 2: html

# output:
#   html_document:
#     theme: united
#     df_print: kable
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
author:
    'Guido Schl√∂gel, Sonja Tockner'
---

```{r setup, include=FALSE}
# keep the options of the original tutorial, hides unnecessary output
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)
```

### Task specification: Case study

In the following code replecates the [seurat tutorial](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html). The code for this notebook can be found [here](https://github.com/gschloegel/Course_Fallstudie_scRNA-Seq-with-Seurat/blob/main/scripts/replicated_tutorial_pbmc3k.rmd)

A dataset of Peripheral Blood Mononuclear Cells(PBMC) is analyzed. The dataset is from 10X Genomics and consists out of 2700 single cells sequenced with Illumina's NextSeq 500. 

### Loading libraries
```{r}
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
```

## Setup the seurat Object
Reading in the data: The Read10X() function is used to read the output of the Cellranger pipeline from 10X. Cellranger is an analysis pipeline that processes single-cell data to align reads, generates feature barcode matrices, performs clustering and other secondary analysis. The output is an UMI(Unique Molecular Identifier) count martix. UMIs are short sequences that are added to DNA sequences to uniquely identify the DNA molecule. The values in the matrix represent the number of molecules for each feature (gene, row) detected in each cell (column). The Count Matrix is used to create the Seurat object, which contains both the data and the analysis, e.g. PCA, clustering results, for a data set. 

```{r}
# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "../data/pbmc3k/filtered_gene_bc_matrices/hg19/")
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
```

## Preprocessing and Quality control
Preprocessing means the selection and filtering of cells based on QC metrices, normalization of the data and scaling or the detection of highly variable features. QC metrices are defined by the user and are, for example: the number of unique genes in each cell (cells with poor quality or empty droplets often contain only a few genes, duplicates or multiplets can have very high gene counts). Unique genes correlate strongly with the number of molecules that are detected within a cell. Another metric is the percentage of reads that map to the mitochondrial genome. Dying and poor quality cells often have extensive mitochondrial contamination. The calculation is done with PercentageFeatureSet (), a function that calculates the percentage of counts from a certain feature set. (all genes starting with MT).
```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

```

## Visualization of QC Metrices
The filter criteria for the cells are defined as follows: Unique feature count > 2500 or < 200 and mitochondrial counts > 5%. 
```{r}
# Visualize QC metrics as a violin plot
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

```

## Visualization of feature-feature relationships by FeatureScatter
```{r}
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```

## Subsets with specific filter criteria can also be defined
```{r}
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)

```

## Normalization of the data
After the filter steps, the data must be normalized. By default, the "LogNormalize" method is used for this, which normalizes the feature expression measurements for each cell by the total expression, multiplies by the scale factor (default 10000) and log-transforms the result. 

```{r}
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)

```

```{r}
### the same
pbmc <- NormalizeData(pbmc)

```

## Feature selection for the identification of highly variable features
A subset of features is defined with high cell-to-cell variation in the expression (highly expressed in some cells and lowly expressed in others); the further focus of the analysis is then placed on these genes. Therefore the FindVariablesFeatures () function is used which allows directy modeling of the mean-variance relationship in single-cell data. Default: 2000 features per dataset, used e.g. for PCA
```{r}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2


```

## Scaling of the data
Scaling is the linear transformation which is the standard before dimension reduction techniques like PCA. (mean expression across cells is 0 and variance across cells is 1). Scaling is intended to achieve equal weighting in downstream analysis so that highly expressed genes do not dominate. 
```{r}

all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
```

## Linear dimension reduction
The next step is to perform PCA on the scaled data. By default, the previously defined features are used, but a new subset can also be defined. 
```{r}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))

```

## Visualizing cells and features that define PCA
```{r}
# Examine and visualize PCA results a few different ways
print(pbmc[["pca"]], dims = 1:5, nfeatures = 5)

```

```{r}
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")

```

```{r}
DimPlot(pbmc, reduction = "pca")

```

## Visualization with DimHeatmap()
DimHeatmap() allows visualization of the primary source of heterogeneity in the dataset and can help to decide which PCs to use for further downstream analysis. Cells and features are ordered by their PCA scores, which also increases speed. This is especially important for large datasets. 
```{r}
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)
```

```{r}
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)

```

## Determine the dimensionality of a dataset
To eliminate technical noise in the individual features, Seurat clusters the cells based on their PCA scores, each PC representing a so-called "metafeature" which combines information about a correlated feature set. In a resampling test, a subset of the data (1% by default) is randomly selected and the PCA is repeated. A null distribution of feature scores is identified...the procedure is repeated...This allows one to identify significant PCs that contain many low p-values. 
```{r}
# NOTE: This process can take a long time for big datasets, comment out for expediency. More
# approximate techniques such as those implemented in ElbowPlot() can be used to reduce
# computation time
pbmc <- JackStraw(pbmc, num.replicate = 100)
pbmc <- ScoreJackStraw(pbmc, dims = 1:20)

```

## Visualization
Comparison of the distribution of p-values for each PC with a unit distribution. Significant PCs show a strong enrichment of features with low p-values (solid curve above the dashed line), we can see a sharp drop-off in significance after the first 10 - 12 PCs.  
```{r}
JackStrawPlot(pbmc, dims = 1:15)

```

## Alternative: Elbow plot
This plot shows a ranking of PCs based on the percentage of variance explained. Cutoff beteen PC 7 -12
```{r}
ElbowPlot(pbmc)

```

### Conclusio: To identify the true dimensionality of a dataset is a challenge. Therefore, several approaches should always be included.
1. Supervised (exploring PCs to determine sources for heterogeneity), e.g. in combination with GSEA
2. Statistical test based on the random null model (time consuming for large datasets, no clear PC cutoff)
3. Heuristic (calculated instantly)

In this case a heuristic approach is used. The dicission is made by looking at the plots. This is subjective as different persons could come to different results and has a negative effect on reproducibility.

We choose 10 PCS for the cutoff, but the following still needs to be considered:
Dendritic cell and NK aficionados may recognize that genes strongly associated with PCs 12 and 13 define rare immune subsets (i.e. MZB1 is a marker for plasmacytoid DCs). However, these groups are so rare, they are difficult to distinguish from background noise for a dataset of this size without prior knowledge.
> repeat downstream analyzes with a different number of PCs (10, 15, or even 50!)


## Cluster the cells
Seutrat uses a graph-based clustering approach. The distance metric remains the same and is based on the previously identified PCs. The approach is based on manuscripts applied to graph-based clustering approaches with scRNA seq data and is characterized by a graph structure, e.g. KNN nearest neighbors. (edges drawn between cells with similar feature expression patterns) and then graph is partitioned into highly interconnected communities. 

First, a KNN graph is constructed based on the Euclidean distance (in PCA space), edge weights are redefined between all two cells based on the overlap in their neighborhood. This step is done using the FindNeighbors() function with the first 10 PCs as input. 
To cluster the cells, modularity optimization techniques such as the Louvain algorithm or SLM are applied. Groups are iteratively clustered with the goal of optimizing the standard modularity function. The FindClusters() function is used therefore, this sets the granularity for downstream clustering (higher values, more clusters). 

Settings: 0.4 - 1.2 for single-cell datasets with about 3K cells. The larger the dataset, the more likely is an optimal solution. With the Idents() function the clusters can be found. 


```{r}
pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)
```

```{r}
# Look at cluster IDs of the first 5 cells
head(Idents(pbmc), 5)

```

## Non-linear dimensional reduction(UMAP)
For this we use UMAC, tSNE is offered as an alternative option.

The goal of nonlinear dimensional reduction techniques is to learn underlying manifold of the data and place similar cells together in low-dimensional space. Cells in the previously determined clusters are to be displayed co-localized in the plots. The same PCs are used as input as for the cluster analysis.

```{r}
pbmc <- RunUMAP(pbmc, dims = 1:10)

```

```{r}
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(pbmc, reduction = "umap")

```

We save the Seurat Object. This can be used to do further analysis without running the time consuming clustering again.

```{r}
### output ordner angelegt
saveRDS(pbmc, file = "../output/pbmc_tutorial.rds")

```

## Search for differentially expressed traits (cluster biomarkers)
Now, markers are to be found which define clusters through differential gene expression. By default, positive and negative markers of a single cluster are identified compared to all other cells. Set threshold value: A trait must be recognized to a minimum percentage in one of the two cell groups and must be expressed differently between the two groups. (both set to 0 leads to dramatic time savings). The maximum number of cells can also be defined (increases the speed)

```{r}
# find all markers of cluster 2
cluster2.markers <- FindMarkers(pbmc, ident.1 = 2, min.pct = 0.25)
head(cluster2.markers, n = 5)

```

```{r}
### find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25)
head(cluster5.markers, n = 5)

```

```{r}
# find markers for every cluster compared to all remaining cells, report only the positive
# ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)

```

## Visualization of the marker expression
The violin plot shows the expression probability distribution over the clusters. The feature plot visualizes the feature expression in a tSNE or PCA plot. RidgePlot (), CellScatter () and DotPlot () can be further options.
```{r}
VlnPlot(pbmc, features = c("MS4A1", "CD79A"))
```

```{r}
# you can plot raw counts as well
VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = "counts", log = TRUE)
```

```{r}
FeaturePlot(pbmc, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP",
    "CD8A"))
```

## DoHeatmap () can be used to generate an expression heatmap for cells and features. The top 20 markers for each cluster are plotted here.
```{r}
pbmc.markers %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC) -> top10
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```

## Assigning cell type identity to clusters
As the last step in the workflow, canonical markers are sent to assign the unbiased clustering to known cell types (9 assignments: "Naive CD4 T", "CD14 + Mono", "Memory CD4 T", "B", "CD8 T", " FCGR3A + Mono ", "NK", "DC", "Platelet"). The new clusters are visualized as umap with DimPlot ().
    
```{r}
new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T", "FCGR3A+ Mono",
    "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

We save this final plot for later use.

```{r save.img, include=FALSE}
library(ggplot2)
plot <- DimPlot(pbmc, reduction = "umap", label = TRUE, label.size = 4.5) + xlab("UMAP 1") + ylab("UMAP 2") + 
  theme(axis.title = element_text(size = 18), legend.text = element_text(size = 18)) + 
  guides(colour = guide_legend(override.aes = list(size = 10)))
ggsave(filename = "../output/pbmc3k_umap.jpg", height = 7, width = 12, plot = plot, quality = 50)
```

Finally we save the final results as Seurat object.

```{r}
saveRDS(pbmc, file = "../output/pbmc3k_final.rds")
```


<details>
  <summary>**Session Info**</summary>
```{r}
sessionInfo()
```
</details>